A generalized framework for accuracy benchmarks

gostevehoward opened the pull request:
This is kind of an unsatisfying PR, because

* `Stripe82Score.jl` and `validate_on_stripe82.jl` can't be deleted yet, because the new versions aren't quite working the same yet, and we can't abandon the old files until the new implementation is satisfactory
* The output of `GalsimBenchmark.jl` is a bit less useful than it was before (but that will change), and I had to comment out some test cases (temporarily), something I swore I wouldn't do anymore, and
* We don't have support yet for synthetic imagery from the Celeste likelihood, one of the main goals.

But I wanted to get a PR out there because it's already been way too long and there are many lines of diff. And the new implementation *is* a lot more flexible than the old one, and hopefully cleaner.

The [wiki page](https://github.com/jeff-regier/Celeste.jl/wiki/Accuracy-benchmarks) should give some context, but feel free to ask questions or request more context.

This implements most of #510.

gostevehoward committed:
start AccuracyBenchmarks module: ground truth catalogs

gostevehoward committed:
script to generate CSV ground truth catalogs

gostevehoward committed:
get galsim_field.py working with catalogs written by write_ground_truth_catalog.csv

This involved quite some changes:
* I modified AccuracyBenchmarks.jl to substantially change the catalog format, so that only columns
  actually used are included (e.g., don't include star parameters for galaxies), and parameters are
  generally in the format/unit used by Celeste (e.g., colors in (natural) log ratio of nMgys, etc)
* I modified galsim_field to compute a suitable WCS to fit the input catalog. This required
  substantial changes to generate_test_images.py
* And some other smaller changes

gostevehoward committed:
round down pixel dimensions; label generated image file based on catalog name

gostevehoward committed:
standalone script to run celeste on synthetic imagery...

...as generated by galsim_field.py, and, in the future, by whatever script generates FITS imagery
from the Celeste model.

This pulls a bunch of logic from GalsimBenchmark.jl into AccuracyBenchmark.jl, organizing and
refactoring along the way.

gostevehoward committed:
move load_primary into AccuracyBenchmark + refactor

the refactoring makes it resemble `load_coadd_catalog`, but it would be annoying to make them
actually share code, since a bunch of field names differ and some of the formats differ slightly. i
think the thing to do might be to have a script that pulls primary catalog and imagery through
SDSSIO into a standalone catalog and image file which matches the format of the synthetic data. in
which case this problem will go away. but then we're juggling the ground truth catalog, the
"initialization" catalog and the imagery for all these cases, and we probably ought to just embed
everything into a single FITS file with mixed table and image HDUs. and i'm putting that off for now
to avoid feature creep.

gostevehoward committed:
move `celeste_to_df` into AccuracyBenchmark and refactor appropriately

gostevehoward committed:
move error scoring code into AccuracyBenchmark + refactor

gostevehoward committed:
make run_celeste_on_synthetic_field.jl write CSV catalog instead of JLD

gostevehoward committed:
script to write a catalog CSV from primary predictions

gostevehoward committed:
finish scoring code, validate common catalog format

gostevehoward committed:
fix bugs, make run_celeste_on_synthetic_field capable on using SDSS imagery

gostevehoward committed:
rename since now this script can run on SDSS (non-synthetic) imagery too

gostevehoward committed:
remove script which has been subsumed by benchmark/accuracy/run_celeste_on_field.jl

gostevehoward committed:
remove stripe82 benchmark scripts, now subsumed by AccuracyBenchmark.jl...

...and benchmark/accuracy/*

gostevehoward committed:
support single inference in run_celeste_on_field

gostevehoward committed:
support initializing celeste with full primary data

to better emulate the previous behavior of validate_on_stripe82.jl

gostevehoward committed:
mount Celeste root dir in GalSim Vagrant VM

this makes it easy to reference catalogs generated under `benchmark/accuracy/output` when running
the GalSim field generation script, `benchmark/galsim/galsim_field.py`

gostevehoward committed:
centralize code for appending hash to output filenames

gostevehoward committed:
no need to print filename; write_catalog() does it

gostevehoward committed:
append hash to predictions as well

will be useful for, say, running MOG vs. FFT celeste, or single vs. joint inference, and comparing
accuracy

gostevehoward committed:
Merge branch 'master' into unify_galsim_stripe82

Resurrecting the master versions of

* benchmark/stripe82/validate_on_stripe82.jl
* src/Stripe82Score.jl
* test/test_score.jl

because the new code's output is differing substantially from the old benchmark results, so there's
work to be done, and we can't just delete the old benchmark in the meantime. But it's already been
too long between pull requests. The duplicate code is regrettable but very much temporary and these
files should be removed as soon as the new benchmarks are up to par.

gostevehoward committed:
work in progress README

gostevehoward committed:
direct support for scoring just one prediction catalog

gostevehoward committed:
finish README for accuracy benchmarks

gostevehoward committed:
minor bugfixes

gostevehoward committed:
remove remaining duplication between GalsimBenchmark and AccuracyBenchmark

gostevehoward committed:
actually i want the README to be in the wiki

gostevehoward committed:
bugfixes

gostevehoward committed:
WIP get galsim benchmarks working

gostevehoward committed:
Merge branch 'master' into unify_galsim_stripe82

gostevehoward committed:
centralize code to convert ref flux + colors to fluxes

gostevehoward committed:
fix conversion of e_scale to half-light radius

gostevehoward committed:
simplify some code

gostevehoward committed:
finish fixing galsim unit tests

we have to comment out some tests, because we're not overriding `min_radius_pix=8.0` in
`load_active_pixels!` now. the way i was doing that before was ugly, and it would be even uglier now
and isn't worth it. we should just start implementing the configurations idea that i proposed in PR

jeff-regier commented:
LGTM. Nice to see it coming together. I appreciate the wiki documentation. Please go ahead and merge this PR whenever the Travis tests are passing.

gostevehoward committed:
forgot to re-add Stripe82Score to Celeste.jl includes

gostevehoward committed:
switch back to old distance formula (with cosd)

gostevehoward committed:
fix handling of NAs in fluxes_from_colors

this little bug was introduced in 13f77ece. The handling of NAs/DataArrays is so awkward :-/.

gostevehoward commented:
For what it's worth, I think things are OK now but data.sdss3.org seems to be down so the travis tests can't run :-/. I'll wait for it to come back up.
